docs/tutorials/typescript/work-queue-slack-app/index.mdx

```diff
@@ -0,0 +1,991 @@
+      // Set the Workflow status to Cancelled by throwing the CancelledFailure error
+      throw e;
+    } else {
+      // Handle other types of errors
+      throw e;
+    }
+  }
+  await continueAsNew<typeof workqueue>(wqdata);
+}
+```
+
+<!--SNIPEND-->
+
+What is also handy about the `condition` API is that if a Cancellation request is made to the Workflow, the condition call will throw a Cancellation error.
+
+In the context of Temporal, the code above will create a Workflow Execution that stays running until it is explicitly told to stop.
+
+### Send data in and out of the Workflow
+
+Next, you will want to define and handle Signals and Queries.
+
+**[Signals](https://docs.temporal.io/encyclopedia/application-message-passing#signals)** are a way to send data into a Workflow.
+**[Queries](https://docs.temporal.io/encyclopedia/application-message-passing#queries)** are a way to read the state of a Workflow.
+
+It is a good practice to define the Signals and Queries outside of the Workflow and handle them right after declaring local state variables within the Workflow.
+
+<!--SNIPSTART typescript-slack-app-temporal-application-workqueue-workflow {"selectedLines": ["1", "6-9", "12-21", "24-49", "66"]}-->
+
+[docs/tutorials/typescript/work-queue-slack-app/code/temporal-application/src/workflows/workqueue.ts](https://github.com/temporalio/temporal-learning/blob/edu-2636-workqueue-slackbot-tut/docs/tutorials/typescript/work-queue-slack-app/code/temporal-application/src/workflows/workqueue.ts)
+
+```ts
+import {
+  // ...
+  defineQuery,
+  defineSignal,
+  setHandler,
+} from "@temporalio/workflow";
+// ...
+export const getWorkqueueDataQuery =
+  defineQuery<WorkqueueData[]>("getWorkqueueData");
+export const addWorkToQueueSignal =
+  defineSignal<[WorkqueueData]>("addWorkqueueData");
+export const claimWorkSignal =
+  defineSignal<[{workId: string; claimantId: string}]>("claimWork");
+export const completeWorkSignal =
+  defineSignal<[{workId: string}]>("completeWork");
+
+export async function workqueue(existingData?: WorkqueueData[]): Promise<void> {
+  // ...
+  // Register a Query handler for 'getWorkqueueData'
+  setHandler(getWorkqueueDataQuery, () => {
+    return wqdata;
+  });
+
+  // Register the Signal handler for adding work
+  setHandler(addWorkToQueueSignal, (data: WorkqueueData) => {
+    wqdata.push(data);
+  });
+
+  // Register Signal handler for claiming work
+  setHandler(claimWorkSignal, ({workId, claimantId}) => {
+    const workItem = wqdata.find((item) => item.id === workId);
+    if (workItem) {
+      workItem.claimantId = claimantId;
+      workItem.status = 2;
+    }
+  });
+
+  // Register Signal handler for completing work
+  setHandler(completeWorkSignal, ({workId}) => {
+    const index = wqdata.findIndex((item) => item.id === workId);
+    if (index !== -1) {
+      wqdata.splice(index, 1);
+    }
+  });
+  // ...
+}
+```
+
+<!--SNIPEND-->
+
+### Test the Workflow with Jest framework
```

Comment:
I'm tempted to suggest moving this out of Step 3 to an optional Step 5 (entirely dedicated to testing) after the more interesting part of the application is finished, *unless* we were to talk about actually mocking up the Slack API calls here so we could see which data is being sent/tested.

Reply:
I like this idea.